### PR TITLE
Sync to llama.cpp + GGML version as of 20230407 09:57 AM UTC.

### LINKED ISSUE #67: Making results independent from threadcount/batch size (from llama.cpp)
This may be something to keep an eye on: https://github.com/ggerganov/llama.cpp/pull/439

Looks like the corresponding code is here: https://github.com/rustformers/llama-rs/blob/bf7bdbcfff3114dcbdafb6eb7eed58f04f19b1c3/llama-rs/src/lib.rs#L1203

According to the comments in the pull, it should trade a small amount of performance for less memory usage. However, at least one user commented they saw more memory use (not sure what size model).

### LINKED ISSUE #118: Update to the latest `llama.cpp` `ggml`
Another day, another suite of changes. Since our last update, M1 and AVX inference has improved, as well as a bunch of other improvements: https://github.com/ggerganov/llama.cpp/compare/437e77855a54e69c86fe03bc501f63d9a3fddb0e..HEAD

If you're interested in tackling this, check out our contributing document: https://github.com/rustformers/llama-rs/blob/main/CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by KerfuffleV2**
This would have been a pretty interesting first issue.

This also includes nuking the `increased_determinism` option since it's no longer needed.

I confirmed that it's possible to change the thread/batch sizes with deterministic results. I haven't don't any extensive performance testing but there didn't seem to be an obvious issue.

Other people should probably do some testing before this gets merged as it's a fairly complicated change. I can't say I really understood what I was porting across.

Closes #118
Closes #67

### CODE REVIEW

File: ggml/src/lib.rs [hunk 1]
```diff
@@ -147,6 +147,12 @@ impl Context {
         self.new_tensor_raw(raw)
     }
 
+    /// Unknown, aside from the obvious. It's transposing something!
```
Comment (philpax) [inline ggml/src/lib.rs:150]: love that documentation from upstream

**Comment by philpax** [comment]
Seems pretty reasonable from what I can see, will need to test but have no objections to merging otherwise

File: ggml/src/lib.rs (hunk 1 above)
Comment (KerfuffleV2) [reply ggml/src/lib.rs:150]: It's not from upstream, it's from me. :)

The original comment was `// ggml_transpose`

Comment (philpax) [reply ggml/src/lib.rs:150]: Yeah that's what I mean - upstream's documentation is... lacking 😅

Comment (KerfuffleV2) [reply ggml/src/lib.rs:150]: Oh, derp. My brain is slightly fried at the moment.

**Comment by philpax** [comment]
Can you also update https://github.com/rustformers/llama-rs/blob/main/ggml-sys/ggml/CREDITS.txt ?

**Comment by KerfuffleV2** [comment]
Can do! By the way, the contributing thing refers to `CREDITS.md`, I didn't know what it meant until now.

**Comment by philpax** [comment]
> Can do! By the way, the contributing thing refers to `CREDITS.md`, I didn't know what it meant until now.

made it clearer, ty

**Review by philpax — APPROVED** [review]
Tested on macOS M1 and Windows x86-64 with no issues. Let's do it 🚀

**Pull request merged**